File: listings/reduce/src/reduce.cpp
#include <cstddef>
#include <cstdlib>
#include <cstdio>
#include <thread>
#include <vector>

constexpr size_t NELEMENTS = 1'000'000;
constexpr size_t NTHREADS = 4;
constexpr size_t QOT = NELEMENTS / NTHREADS;
constexpr size_t REM = NELEMENTS % NTHREADS;

auto main() -> int {
    std::vector<double> vector(NELEMENTS, 1.0);
    std::vector<std::thread> threads(NTHREADS);

    double result = 0.0;
    for (size_t t = 0; t < NTHREADS; ++t) {
        size_t const start = t * QOT;
        size_t const end = t == NTHREADS - 1 ? start + QOT + REM : start + QOT;
        threads[t] = std::thread([&]() {
            double partial_sum = 0.0;
            for (size_t i = start; i < end; ++i) {
                partial_sum += vector[i];
            }
            result += partial_sum;
        });
    }

    for (auto& t: threads) {
        t.join();
    }

    printf("result: %lf\n", result);
    return 0;
}

File: listings/reduce/src/reduce.rs
const NELEMENTS: usize = 1_000_000;
const NTHREADS: usize = 4;
const QOT: usize = NELEMENTS / NTHREADS;
const REM: usize = NELEMENTS % NTHREADS;

fn main() {
    let array = vec![1.0; NELEMENTS];
    let mut threads = Vec::with_capacity(NTHREADS);
    let mut start = 0;

    let mut result = 0.0;
    for t in 0..NTHREADS {
        let loc = if t < REM { QOT + 1 } else { QOT };
        let end = start + loc;
        threads.push(std::thread::spawn(|| {
            let mut partial_sum = 0.0;
            for i in start..end {
                partial_sum += array[i];
            }
            result += partial_sum;
        }));
        start = end;
    }

    for t in threads {
        t.join().unwrap();
    }

    println!("result: {result}");
}
